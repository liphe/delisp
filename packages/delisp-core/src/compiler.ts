import runtime from "@delisp/runtime";
import {
  Expression,
  isDefinition,
  isExport,
  isTypeAlias,
  Module,
  SConditional,
  SDefinition,
  SExport,
  SFunction,
  SFunctionCall,
  SIdentifier,
  SLet,
  SRecord,
  SVectorConstructor,
  Syntax
} from "./syntax";

import { methodCall } from "./compiler/estree-utils";
import { last, mapObject, maybeMap } from "./utils";

import { InvariantViolation } from "./invariant";

import { printHighlightedExpr } from "./error-report";

import {
  DefinitionBackend,
  dynamicDefinition,
  staticDefinition
} from "./compiler/definitions";
import { cjs, esm, ModuleBackend } from "./compiler/modules";

import { member } from "./compiler/estree-utils";
import {
  compileInlinePrimitive,
  isInlinePrimitive
} from "./compiler/inline-primitives";
import { identifierToJS, isValidJSIdentifierName } from "./compiler/jsvariable";
import { pprint } from "./printer";

import * as escodegen from "escodegen";
import * as JS from "estree";

import createDebug from "debug";
const debug = createDebug("delisp:compiler");

interface EnvironmentBinding {
  jsname: string;
  source: "lexical" | "module" | "primitive";
}

export interface Environment {
  defs: DefinitionBackend;
  moduleFormat: ModuleBackend;
  bindings: {
    [symbol: string]: EnvironmentBinding;
  };
}

export interface CompilerOptions {
  definitionContainer?: string;
  esModule?: boolean;
}

function addBinding(varName: string, env: Environment): Environment {
  return {
    ...env,
    bindings: {
      ...env.bindings,
      [varName]: { jsname: identifierToJS(varName), source: "lexical" }
    }
  };
}

function lookupBinding(varName: string, env: Environment) {
  return env.bindings[varName];
}

function compileBody(body: Expression[], env: Environment): JS.BlockStatement {
  const middleForms = body.slice(0, -1);
  const lastForm = last(body)!;

  return {
    type: "BlockStatement",
    body: [
      ...middleForms.map(
        (e): JS.ExpressionStatement => ({
          type: "ExpressionStatement",
          expression: compile(e, env)
        })
      ),
      {
        type: "ReturnStatement",
        argument: compile(lastForm, env)
      }
    ]
  };
}

// Convert a Delisp variable name to Javascript. This function should
// be injective so there is no collisions and the output should be a
// valid variable name.

function compileLambda(
  fn: SFunction,
  env: Environment
): JS.ArrowFunctionExpression {
  const newEnv = fn.lambdaList.positionalArgs.reduce(
    (e, param) => addBinding(param.name, e),
    env
  );

  const jsargs = fn.lambdaList.positionalArgs.map(
    (param): JS.Pattern => ({
      type: "Identifier",
      name: lookupBinding(param.name, newEnv).jsname
    })
  );

  const implicitReturn = fn.body.length === 1;

  const body: JS.Expression | JS.Statement = implicitReturn
    ? compile(fn.body[0], newEnv)
    : compileBody(fn.body, newEnv);

  return {
    type: "ArrowFunctionExpression",
    params: [...jsargs],
    body,
    expression: implicitReturn
  };
}

function compileDefinition(def: SDefinition, env: Environment): JS.Statement {
  const value = compile(def.value, env);
  const name = lookupBinding(def.variable, env).jsname;
  return env.defs.define(name, value);
}

function compileFunctionCall(
  funcall: SFunctionCall,
  env: Environment
): JS.Expression {
  const compiledArgs = funcall.args.map(arg => compile(arg, env));
  if (funcall.fn.tag === "identifier" && isInlinePrimitive(funcall.fn.name)) {
    return compileInlinePrimitive(funcall.fn.name, compiledArgs, "funcall");
  } else {
    return {
      type: "CallExpression",
      callee: compile(funcall.fn, env),
      arguments: compiledArgs
    };
  }
}

function compileIdentifier(ref: SIdentifier, env: Environment): JS.Expression {
  const binding = lookupBinding(ref.name, env);

  if (!binding) {
    if (isInlinePrimitive(ref.name)) {
      return compileInlinePrimitive(ref.name, [], "value");
    } else {
      return env.defs.access(identifierToJS(ref.name));
    }
  }

  switch (binding.source) {
    case "primitive":
      return member(
        {
          type: "Identifier",
          name: "env"
        },
        binding.jsname
      );
    case "module":
      return env.defs.access(binding.jsname);
    case "lexical":
      return {
        type: "Identifier",
        name: binding.jsname
      };
    default:
      throw new InvariantViolation(
        "This switch-statement should be exhaustive but TS doesn't detect it somehow."
      );
  }
}

function compileConditional(
  expr: SConditional,
  env: Environment
): JS.Expression {
  return {
    type: "ConditionalExpression",
    test: compile(expr.condition, env),
    consequent: compile(expr.consequent, env),
    alternate: compile(expr.alternative, env)
  };
}

function compileLetBindings(expr: SLet, env: Environment): JS.Expression {
  const newenv = expr.bindings.reduce(
    (e, binding) => addBinding(binding.var, e),
    env
  );

  return {
    type: "CallExpression",
    callee: {
      type: "FunctionExpression",
      params: expr.bindings.map(
        (b): JS.Pattern => ({
          type: "Identifier",
          name: lookupBinding(b.var, newenv).jsname
        })
      ),
      body: compileBody(expr.body, newenv)
    },
    arguments: expr.bindings.map(b => compile(b.value, env))
  };
}

function literal(value: number | string): JS.Literal {
  return {
    type: "Literal",
    value
  };
}

function identifier(name: string): JS.Identifier {
  return {
    type: "Identifier",
    name
  };
}

function compileVector(
  expr: SVectorConstructor,
  env: Environment
): JS.Expression {
  return {
    type: "ArrayExpression",
    elements: expr.values.map(e => compile(e, env))
  };
}

function compileRecord(expr: SRecord, env: Environment): JS.Expression {
  const newObj: JS.ObjectExpression = {
    type: "ObjectExpression",
    properties: expr.fields.map(
      ({ label, value }): JS.Property => {
        if (!label.startsWith(":")) {
          throw new InvariantViolation(`Invalid record ${label}`);
        }

        const name = label.replace(/^:/, "");

        return {
          type: "Property",
          key: isValidJSIdentifierName(name) ? identifier(name) : literal(name),
          value: compile(value, env),
          kind: "init",
          method: false,
          shorthand: false,
          computed: false
        };
      }
    )
  };
  if (expr.extends) {
    return methodCall({ type: "Identifier", name: "Object" }, "assign", [
      { type: "ObjectExpression", properties: [] },
      compile(expr.extends, env),
      newObj
    ]);
  } else {
    return newObj;
  }
}

function compileNumber(value: number): JS.Expression {
  if (value >= 0) {
    return literal(value);
  } else {
    return {
      type: "UnaryExpression",
      operator: "-",
      prefix: true,
      argument: literal(-value)
    };
  }
}

export function compile(expr: Expression, env: Environment): JS.Expression {
  switch (expr.tag) {
    case "number":
      return compileNumber(expr.value);
    case "string":
      return literal(expr.value);
    case "vector":
      return compileVector(expr, env);
    case "record":
      return compileRecord(expr, env);
    case "identifier":
      return compileIdentifier(expr, env);
    case "conditional":
      return compileConditional(expr, env);
    case "function":
      return compileLambda(expr, env);
    case "function-call":
      return compileFunctionCall(expr, env);
    case "let-bindings":
      return compileLetBindings(expr, env);
    case "type-annotation":
      return compile(expr.value, env);
  }
}

function compileTopLevel(
  syntax: Syntax,
  env: Environment
): JS.Statement | null {
  if (isExport(syntax) || isTypeAlias(syntax)) {
    // exports are compiled at the end of the module
    return null;
  }

  let js: JS.Statement;

  if (isDefinition(syntax)) {
    js = compileDefinition(syntax, env);
  } else {
    js = {
      type: "ExpressionStatement",
      expression: compile(syntax, env)
    };
  }

  return {
    ...js,
    // Include a comment with the original source code immediately
    // before each toplevel compilation.
    leadingComments: [
      {
        type: "Block",
        value: `
${pprint(syntax, 60)}
`
      }
    ]
  };
}

function compileRuntime(env: Environment): JS.Statement | JS.ModuleDeclaration {
  return env.moduleFormat.importRuntime("env");
}

function compileExports(
  exps: SExport[],
  env: Environment
): Array<JS.Statement | JS.ModuleDeclaration> {
  const exportNames = exps.map(exp => {
    const binding = lookupBinding(exp.value.name, env);
    if (!binding || binding.source !== "module") {
      throw new Error(
        printHighlightedExpr(
          "You can only export user definitions",
          exp.value.location
        )
      );
    } else {
      return binding.jsname;
    }
  });

  if (exportNames.length > 0) {
    return [env.moduleFormat.export(exportNames)];
  } else {
    return [];
  }
}

export function moduleEnvironment(
  m: Module,
  opts: CompilerOptions = {}
): Environment {
  const moduleDefinitions = m.body
    .filter(isDefinition)
    .map(decl => decl.variable);
  const moduleBindings = moduleDefinitions.reduce(
    (d, decl) => ({
      ...d,
      [decl]: { jsname: identifierToJS(decl), source: "module" }
    }),
    {}
  );

  const primitiveBindings = mapObject(
    runtime,
    (_, key: string): EnvironmentBinding => ({
      jsname: key,
      source: "primitive"
    })
  );

  const initialEnv = {
    defs: opts.definitionContainer
      ? dynamicDefinition(opts.definitionContainer)
      : staticDefinition,
    moduleFormat: opts.esModule ? esm : cjs,
    bindings: { ...primitiveBindings, ...moduleBindings }
  };

  return initialEnv;
}

function compileModule(
  m: Module,
  includeRuntime: boolean,
  env: Environment
): JS.Program {
  return {
    type: "Program",
    sourceType: "module",
    body: [
      ...(includeRuntime ? [compileRuntime(env)] : []),
      ...maybeMap((syntax: Syntax) => compileTopLevel(syntax, env), m.body),
      ...compileExports(m.body.filter(isExport), env)
    ]
  };
}

export function compileToString(syntax: Syntax, env: Environment): string {
  const ast = compileModule({ tag: "module", body: [syntax] }, false, env);
  const code = escodegen.generate(ast, { comment: true });
  debug("jscode:", code);
  return code;
}

export function compileModuleToString(
  m: Module,
  opts: CompilerOptions = {}
): string {
  const env = moduleEnvironment(m, opts);
  const ast = compileModule(m, true, env);
  const code = escodegen.generate(ast, { comment: true });
  debug("jscode:", code);
  return code;
}
