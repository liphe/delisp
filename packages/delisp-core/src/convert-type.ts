import { capitalize } from "./utils";
import { ConvertError, parseRecord } from "./convert-utils";
import { printHighlightedExpr } from "./error-report";
import { readFromString } from "./reader";
import { last } from "./utils";

import {
  isSymbolOfName,
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";

import {
  emptyRow,
  TConstant,
  TVar,
  Type,
  tApp,
  tBoolean,
  tNumber,
  tRecord,
  tMultiValuedFunction,
  tValues,
  tCases,
  tEffect,
  tString,
  tUserDefined,
  tVar,
  tVector,
  tVoid,
  tNone,
  tcArrow,
  tcStar,
  TypeSchema
} from "./types";

import { TypeWithWildcards } from "./type-wildcards";

/** Return true if a symbol is a valid name for a user defined type, false otherwise. */
export function userDefinedType(expr: ASExprSymbol): boolean {
  return /^[A-Z]/.test(expr.name);
}

/** Check if a symbol is a valid user defined type name or throw a user-friendly error otherwise. */
export function checkUserDefinedTypeName(expr: ASExprSymbol): void {
  if (!userDefinedType(expr)) {
    throw new ConvertError(
      printHighlightedExpr(
        `'${
          expr.name
        }' is not a valid name for a type. Type names should start with a capital letter. Try '${capitalize(
          expr.name
        )}'?`,
        expr.location
      )
    );
  }
}

function convertSymbol(expr: ASExprSymbol): TVar | TConstant {
  switch (expr.name) {
    case "->":
      return tcArrow;
    case "boolean":
      return tBoolean;
    case "number":
      return tNumber;
    case "string":
      return tString;
    case "void":
      return tVoid;
    case "none":
      return tNone;
    case "*":
      return tcStar;
    default:
      return userDefinedType(expr) ? tUserDefined(expr.name) : tVar(expr.name);
  }
}

function convertEffect(effects: ASExpr[]): Type {
  const labels = effects.map(e => {
    if (e.tag !== "symbol") {
      throw new ConvertError(
        printHighlightedExpr(`not a valid effect`, e.location)
      );
    }
    return e.name;
  });
  const last = labels.slice(-2);
  if (last.length === 2 && last[0] === "|") {
    return tEffect(labels.slice(0, -2), tVar(last[1]));
  } else {
    return tEffect(labels);
  }
}

function convertCases(_op: ASExpr, args: ASExpr[]): Type {
  const lastArg = last(args);
  let tail: ASExpr | undefined;

  if (lastArg && lastArg.tag === "symbol" && !lastArg.name.startsWith(":")) {
    tail = args.pop();
  }

  function parseCase(c: ASExpr): { label: string; type: Type } {
    if (c.tag === "symbol" && c.name.startsWith(":")) {
      return { label: c.name, type: tVoid };
    } else if (c.tag === "list") {
      if (c.elements.length !== 2) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected exaclty two elements`,
            c.location
          )
        );
      }
      const [keyword, innerType] = c.elements;
      if (keyword.tag !== "symbol" || !keyword.name.startsWith(":")) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected valid keyword`,
            keyword.location
          )
        );
      }
      return { label: keyword.name, type: convert_(innerType) };
    } else {
      throw new ConvertError(printHighlightedExpr(`Invalid case`, c.location));
    }
  }

  return tCases(args.map(parseCase), tail && convert_(tail));
}

function convertValues(expr: ASExpr): Type[] {
  // Check if X is the symbol VALUES
  if (
    expr.tag === "list" &&
    expr.elements.length > 0 &&
    isSymbolOfName(expr.elements[0], "values")
  ) {
    return expr.elements.slice(1).map(convert_);
  } else {
    return [convert_(expr)];
  }
}

function convertList(expr: ASExprList): Type {
  const [op, ...args] = expr.elements;
  if (op.tag === "symbol" && op.name === "effect") {
    return convertEffect(args);
  } else if (op.tag === "symbol" && op.name === "cases") {
    return convertCases(op, args);
  } else if (op.tag === "symbol" && op.name === "->") {
    if (args.length < 2) {
      throw new ConvertError(`Missing some arguments!`);
    }
    const fnargs = args.slice(0, -2);
    const fneffect = args[args.length - 2];
    const fnout = args[args.length - 1];
    return tMultiValuedFunction(
      fnargs.map(convert_),
      convert_(fneffect),
      tValues(convertValues(fnout))
    );
  } else {
    const opType = convert_(op);
    return tApp(opType, ...args.map(convert_));
  }
}

function convertVector(expr: ASExprVector): Type {
  if (expr.elements.length !== 1) {
    throw new ConvertError(
      printHighlightedExpr("Expected exactly 1 argument", expr.location)
    );
  }
  return tVector(convert_(expr.elements[0]));
}

function convertMap(expr: ASExprMap): Type {
  const { fields, tail } = parseRecord(expr);

  return tRecord(
    fields.map(({ label, value }) => ({
      label: label.name,
      type: convert_(value)
    })),
    tail ? convert_(tail) : emptyRow
  );
}

/* Try to convert a S-Expression into a type. */
function convert_(expr: ASExpr): Type {
  switch (expr.tag) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
    default:
      throw new ConvertError(
        printHighlightedExpr("Not a valid type", expr.location)
      );
  }
}

function convertToTypeWithWildcards(expr: ASExpr): TypeWithWildcards {
  return new TypeWithWildcards(convert_(expr));
}

export { convertToTypeWithWildcards as convert };

export function readType(source: string): TypeSchema {
  const t = convertToTypeWithWildcards(readFromString(source));
  return t.generalize();
}
