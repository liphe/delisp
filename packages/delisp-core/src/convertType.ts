import { printHighlightedExpr } from "./error-report";
import { ASExpr, ASExprList, ASExprSymbol } from "./sexpr";
import { Type } from "./types";

function convertSymbol(expr: ASExprSymbol): Type {
  switch (expr.name) {
    case "number":
      return { type: "number" };
    case "string":
      return { type: "string" };
    default:
      return { type: "type-variable", name: expr.name };
  }
}

function convertList(expr: ASExprList): Type {
  const [first, from, to] = expr.elements;

  if (first.type !== "symbol" || first.name !== "->") {
    throw new Error(printHighlightedExpr("Expected -> as first symbol", expr));
  }

  if (expr.elements.length !== 3) {
    throw new Error(printHighlightedExpr("Expected 3 elements", expr));
  }

  if (from.type !== "list") {
    throw new Error(
      printHighlightedExpr("Expected second element to be a list", expr)
    );
  }

  return {
    type: "function",
    from: from.elements.map(convert),
    to: convert(to)
  };
}

export function convert(expr: ASExpr): Type {
  switch (expr.type) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    default:
      throw new Error(printHighlightedExpr("Not a valid type", expr));
  }
}
