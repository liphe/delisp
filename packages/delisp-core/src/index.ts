import { convert as convertSyntax } from "./convert";

import { readFromString } from "./reader";

export { collectConvertErrors } from "./convert";

export {
  compileToString,
  compileModuleToString,
  moduleEnvironment
} from "./compiler";
export { createContext, evaluate, evaluateModule } from "./eval";
export {
  inferType,
  inferModule,
  inferExpressionInModule,
  defaultEnvironment
} from "./infer";
export { Type } from "./types";
export { printType } from "./type-printer";
export {
  isDeclaration,
  isDefinition,
  isExpression,
  isExport,
  isTypeAlias
} from "./syntax";
export { default as primitives } from "./primitives";

export { printHighlightedExpr } from "./error-report";

export function readSyntax(source: string) {
  return convertSyntax(readFromString(source));
}

export { findSyntaxByOffset, findSyntaxByRange } from "./syntax-utils";

export {
  createModule,
  readModule,
  addToModule,
  removeModuleDefinition,
  removeModuleTypeDefinition,
  moduleDefinitions,
  moduleDefinitionByName
} from "./module";

export { generateTSModuleDeclaration } from "./typescript-generation";

export { Encoder } from "./prettier";
export { pprintAs, pprint, pprintModuleAs, pprintModule } from "./printer";
