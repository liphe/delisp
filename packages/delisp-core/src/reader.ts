//
// reader.ts --- The reader for Deisp
//

//
// S-Expression parser
//

import {
  alternatives,
  atLeastOne,
  character,
  delimited,
  delimitedMany,
  endOfInput,
  many,
  Parser,
  until
} from "./parser-combinators";

import { ASExpr } from "./sexpr";

const spaces = many(
  alternatives(character(" "), character("\t"), character("\n"))
);

function spaced<A>(x: Parser<A>) {
  return delimited(spaces, x, spaces);
}

const alphanumeric = character().chain(char => {
  if (/[a-zA-Z0-9]/.test(char)) {
    return Parser.of(char);
  } else {
    return Parser.fail("an alphanumeric character");
  }
});

const digit = character().chain(char => {
  if (/[0-9]/.test(char)) {
    return Parser.of(char);
  } else {
    return Parser.fail("a digit");
  }
});

//
// Number
//

const unsignedInteger: Parser<number> = atLeastOne(digit).map(digits =>
  parseInt(digits.join(""), 10)
);
const negativeInteger: Parser<number> = character("-")
  .chain(_ => unsignedInteger)
  .map(num => -num);

const numberP: Parser<ASExpr> = alternatives(negativeInteger, unsignedInteger)
  .map(
    (val, location): ASExpr => ({
      type: "number",
      value: val,
      location
    })
  )
  .description("number");

//
// String
//
const doubleQuote = character('"').description("double quote");

const stringChar = character()
  .chain(char => {
    if (char === '"') {
      return Parser.fail("Not inside string");
    }
    if (char === "\\") {
      return alternatives(
        character("n").map(_ => "\n"),
        character("\\")
      ).description("escaped string character");
    }
    return Parser.of(char);
  })
  .description("string character");

const stringP = delimitedMany(doubleQuote, stringChar, doubleQuote)
  .map(
    (chars, location): ASExpr => ({
      type: "string",
      value: chars.join(""),
      location
    })
  )
  .description("string");

//
// Symbol
//

const symbol: Parser<ASExpr> = atLeastOne(
  alternatives(
    alphanumeric,
    character("!"),
    character("@"),
    character("#"),
    character("$"),
    character("%"),
    character("^"),
    character("&"),
    character("*"),
    character("<"),
    character(">"),
    character("+"),
    character("-"),
    character("/"),
    character("~"),
    character("?")
  )
)
  .map(
    (chars, location): ASExpr => ({
      type: "symbol",
      name: chars.join(""),
      location
    })
  )
  .description("symbol");

//
// Lists & S-Expressions
//

const atom: Parser<ASExpr> = alternatives(numberP, stringP, symbol).description(
  "atom"
);

const leftParen = character("(").description("open parenthesis");
const rightParen = character(")").description("close parenthesis");

function list(x: Parser<ASExpr>): Parser<ASExpr> {
  return delimitedMany(leftParen, x, spaces.then(rightParen))
    .map(
      (elements, location): ASExpr => ({
        type: "list",
        elements,
        location
      })
    )
    .description("list");
}

const reportUnmatched: Parser<{}> = Parser.lookahead(rightParen).chain(
  closed => {
    if (closed) {
      return Parser.fail("Unmatched closed parenthesis");
    } else {
      return Parser.of({});
    }
  }
);

const sexpr: Parser<ASExpr> = spaced(
  reportUnmatched.then(atom.or(() => list(sexpr)))
);

const sexprs: Parser<ASExpr[]> = until(endOfInput, sexpr);

export function readAllFromString(str: string): ASExpr[] {
  return sexprs.parse(str);
}

//
// Parser a Delisp expression from a string
//
export function readFromString(str: string): ASExpr {
  return sexpr.parse(str);
}
