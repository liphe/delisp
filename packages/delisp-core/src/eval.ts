//
// Evaluation
//

import * as vm from "vm";
import { compileToString, Environment, moduleEnvironment } from "./compiler";
import primitives from "./primitives";
import { Module, Syntax } from "./syntax";
import { mapObject } from "./utils";

import {
  caseTag,
  matchTag,
  pair,
  fst,
  snd,
  values,
  primaryValue,
  bindPrimaryValue
} from "@delisp/runtime";

export function createContext() {
  const sandbox = {
    env: mapObject(primitives, p => p.value),
    console,
    // Primitives
    caseTag,
    matchTag,
    pair,
    fst,
    snd,
    values,
    primaryValue,
    bindPrimaryValue
  };
  vm.createContext(sandbox);
  return sandbox;
}

export function evaluate(
  syntax: Syntax,
  env: Environment,
  context = createContext()
): unknown {
  const code = compileToString(syntax, env);
  const result = vm.runInContext(code, context);
  return result;
}

export function evaluateModule(m: Module, context = createContext()): void {
  const env = moduleEnvironment(m, { definitionContainer: "env" });
  m.body.forEach(s => {
    return evaluate(s, env, context);
  });
}
