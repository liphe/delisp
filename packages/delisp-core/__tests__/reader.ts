import { readAllFromString, readFromString } from "../src/reader";

describe("Reader", () => {
  it("should read numbers", () => {
    expect(readFromString("12")).toMatchObject({
      type: "number",
      value: 12,
      location: { start: 0, end: 2 }
    });
    expect(readFromString("  12  ")).toMatchObject({
      type: "number",
      value: 12,
      location: { start: 2, end: 4 }
    });
    expect(readFromString("  -12  ")).toMatchObject({
      type: "number",
      value: -12,
      location: { start: 2, end: 5 }
    });
  });

  it("should read strings", () => {
    expect(readFromString('  "xyz"  ')).toMatchObject({
      type: "string",
      value: "xyz",
      location: { start: 2, end: 7 }
    });

    expect(readFromString('  "a\\nb"  ')).toMatchObject({
      type: "string",
      value: "a\nb",
      location: { start: 2, end: 8 }
    });
  });

  it("should read symbols", () => {
    expect(readFromString("  xyz  ")).toMatchObject({
      type: "symbol",
      name: "xyz",
      location: { start: 2, end: 5 }
    });

    expect(readFromString("  a2  ")).toMatchObject({
      type: "symbol",
      name: "a2",
      location: { start: 2, end: 4 }
    });

    expect(readFromString("  $bc  ")).toMatchObject({
      type: "symbol",
      name: "$bc",
      location: { start: 2, end: 5 }
    });
  });

  it("should read lists", () => {
    expect(readFromString("()")).toMatchObject({
      type: "list",
      shape: "round",
      elements: [],
      location: { start: 0, end: 2 }
    });
    expect(readFromString("(  )")).toMatchObject({
      type: "list",
      shape: "round",
      elements: [],
      location: { start: 0, end: 4 }
    });
    expect(readFromString("(1 2 3)")).toMatchObject({
      type: "list",
      shape: "round",
      elements: [
        {
          type: "number",
          value: 1,
          location: { start: 1, end: 2 }
        },
        {
          type: "number",
          value: 2,
          location: { start: 3, end: 4 }
        },
        {
          type: "number",
          value: 3,
          location: { start: 5, end: 6 }
        }
      ],
      location: { start: 0, end: 7 }
    });

    expect(readFromString(" (1 ( 2 ) 3) ")).toMatchObject({
      type: "list",
      shape: "round",
      elements: [
        {
          type: "number",
          value: 1,
          location: { start: 2, end: 3 }
        },
        {
          type: "list",
          elements: [
            {
              type: "number",
              value: 2,
              location: { start: 6, end: 7 }
            }
          ],
          location: { start: 4, end: 9 }
        },
        {
          type: "number",
          value: 3,
          location: { start: 10, end: 11 }
        }
      ],
      location: { start: 1, end: 12 }
    });
  });

  it("should read lists with square bracket notation", () => {
    expect(readFromString("[]")).toMatchObject({
      type: "list",
      shape: "square",
      elements: [],
      location: { start: 0, end: 2 }
    });
    expect(readFromString("[  ]")).toMatchObject({
      type: "list",
      shape: "square",
      elements: [],
      location: { start: 0, end: 4 }
    });
    expect(readFromString("[1 2 3]")).toMatchObject({
      type: "list",
      shape: "square",
      elements: [
        {
          type: "number",
          value: 1,
          location: { start: 1, end: 2 }
        },
        {
          type: "number",
          value: 2,
          location: { start: 3, end: 4 }
        },
        {
          type: "number",
          value: 3,
          location: { start: 5, end: 6 }
        }
      ],
      location: { start: 0, end: 7 }
    });

    expect(readFromString(" [1 [ 2 ] 3] ")).toMatchObject({
      type: "list",
      shape: "square",
      elements: [
        {
          type: "number",
          value: 1,
          location: { start: 2, end: 3 }
        },
        {
          type: "list",
          elements: [
            {
              type: "number",
              value: 2,
              location: { start: 6, end: 7 }
            }
          ],
          location: { start: 4, end: 9 }
        },
        {
          type: "number",
          value: 3,
          location: { start: 10, end: 11 }
        }
      ],
      location: { start: 1, end: 12 }
    });
  });

  it("should read multiple S-expressions", () => {
    expect(readAllFromString("(x 1 2)(y 3)")).toMatchObject([
      { type: "list" },
      { type: "list" }
    ]);

    expect(readAllFromString(" (x 1 2) (y 3) ")).toMatchObject([
      { type: "list" },
      { type: "list" }
    ]);

    expect(
      readAllFromString(`
      (x 1 2)
      (y 3)
    `)
    ).toMatchObject([{ type: "list" }, { type: "list" }]);
  });

  describe("Error messages", () => {
    const failedRead = (x: string) => {
      try {
        readFromString(x);
        throw new Error(`This expression didn't fail!`);
      } catch (err) {
        return err.message;
      }
    };

    it("generate user-friendly error for an incomplete list", () => {
      expect(failedRead("(1 2 3")).toMatchSnapshot();
    });

    it("generate user-friendly error for a closing parenthesis", () => {
      expect(failedRead(")")).toMatchSnapshot();
    });

    it("generate user-friendly error for an incomplete string", () => {
      expect(failedRead('"foo')).toMatchSnapshot();
    });

    it("generate user-friendly error for badly escaped string", () => {
      expect(failedRead('"ab\\xyz"')).toMatchSnapshot();
    });

    it("generate a user-friendly error for incomplete escaped string", () => {
      expect(failedRead('"abc\\')).toMatchSnapshot();
    });
  });

  it("should detect incomplete inputs", () => {
    const read = (x: string) => {
      try {
        readFromString(x);
        return undefined;
      } catch (err) {
        return err.incomplete;
      }
    };

    const readAll = (x: string) => {
      try {
        readAllFromString(x);
        return undefined;
      } catch (err) {
        return err.incomplete;
      }
    };

    expect(read("(1 2 3")).toBe(true);
    expect(read(")")).toBe(false);
    expect(read('"foo')).toBe(true);
    expect(read('"ab\\xyz"')).toBe(false);
    expect(read('"abc\\')).toBe(true);

    expect(readAll("(1 2 3)(4 5")).toBe(true);
    expect(readAll("(1 2 3)4 5)")).toBe(false);
    expect(readAll("((1 2 3)")).toBe(true);
    expect(readAll("(1 2 3))")).toBe(false);
  });
});
