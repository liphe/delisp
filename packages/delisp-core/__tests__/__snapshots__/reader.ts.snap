// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reader Error messages generate a user-friendly error for incomplete escaped string 1`] = `
"file:1:5: Expected escaped string character
\\"abc\\\\
-----^"
`;

exports[`Reader Error messages generate user-friendly error for a closing parenthesis 1`] = `
"file:1:0: Expected Unmatched closed parenthesis
)
^"
`;

exports[`Reader Error messages generate user-friendly error for an incomplete list 1`] = `
"file:1:6: Expected close parenthesis, number, double quote, symbol, open parenthesis, or open square bracket
(1 2 3
------^"
`;

exports[`Reader Error messages generate user-friendly error for an incomplete string 1`] = `
"file:1:4: Expected double quote, or string character
\\"foo
----^"
`;

exports[`Reader Error messages generate user-friendly error for badly escaped string 1`] = `
"file:1:4: Expected escaped string character
\\"ab\\\\xyz\\"
----^"
`;
